// bbs-utils/lib/route.js
var crypto = require('crypto');
var fs = require('fs');

var routeUtils = {
    md5Encryption: function (str) {
        let MD5_SUFFIX = 's5w84&&d423sda#$@2025s5*4s2';
        let obj1 = crypto.createHash('md5');
        let obj2 = crypto.createHash('md5');
        obj1.update(str + MD5_SUFFIX);
        let md5Value = obj1.digest('hex');
        obj2.update(md5Value + MD5_SUFFIX);
        return obj2.digest('hex');
    },
    getFormatTime: function () {
        let date = new Date();
        let time = date.getFullYear() + "-" + (date.getMonth() + 1) + "-" + date.getDate();
        time +=" " + date.getHours() + ":" + date.getMinutes() + ":" + date.getSeconds();
        return time;
    },
    mylog: function (content) {
        content = this.getFormatTime() + " >> " + content + "\n";
        fs.writeFile('sever.log', content, { encoding: 'utf-8', flag: 'a' }, (erro) => {});
    },
    /**
     * tip的常见配置
     * @param {String} name tips name
     */
    simpleTip: function (name) {
        switch (name) {
            case "login":
                return {
                    title: "未登陆!",
                    time: 3,
                    content: "此处游客无法访问",
                    webName: "登陆页",
                    redirect: "/game/user/login"
                };
            case "auth":
                return {
                    title: "403!",
                    time: 3,
                    content: "权限不足，禁止访问",
                    webName: "游戏首页",
                    redirect: "/game"
                };
            case "404":
                return {
                    title: "404!",
                    time: 3,
                    content: "我还没想好这里放啥",
                    webName: "游戏首页",
                    redirect: "/game"
                };
            case "403":
                return {
                    title: "403!",
                    time: 3,
                    content: "权限不足，禁止访问",
                    webName: "游戏首页",
                    redirect: "/game"
                };
            case "loginsuccess":
                return {
                    title: "登录成功!",
                    time: 2,
                    content: "点击前往页面",
                    webName: "游戏首页",
                    redirect: "/game"
                };
            case "loginfailed":
                return {
                    title: "登录失败!",
                    time: 2,
                    content: "用户名或密码错误",
                    webName: "登录页",
                    redirect: "/game/user/login"
                };
            case "registersuccess":
                return {
                    title: "注册成功!",
                    time: 3,
                    content: "点击前往登录页",
                    webName: "登录页",
                    redirect: "/game/user/login"
                };
            case "registerfailed":
                return {
                    title: "注册失败!",
                    time: 3,
                    content: "用户名已存在",
                    webName: "注册页",
                    redirect: "/game/user/register"
                };
            case "namenotexists":
                return {
                    title: "用户不存在!",
                    time: 3,
                    content: "用户不存在或被封禁",
                    webName: "登录页",
                    redirect: "/game/user/login"
                };
            case "logout":
                return {
                    title: "注销成功!",
                    time: 3,
                    content: "点击前往页",
                    webName: "游戏首页",
                    redirect: "/game"
                };
            default:
                return {
                    title: "未知错误!",
                    time: 3,
                    content: "看看你干了啥!",
                    webName: "游戏首页",
                    redirect: "/game"
                };
        }
    }
};

routeUtils.mylog.bind(routeUtils);

module.exports = routeUtils;
