// bbs-utils/lib/sqlquery.js
const mysql = require('mysql');

// 配置
let options = {
    host: "localhost",
    // port: 3306, 是默认端口可以不用写
    user: "root",
    password: "123456789",
    database: "gamebbs"
};


// 创建连接对象
let connection = mysql.createConnection(options);

// 启动连接
connection.connect((err) => {
    if (err) {
        console.log('--------数据库连接失败--------');
        console.log(err);
    } else {
        console.log('--------数据库连接成功--------');
    }
});

/**
 * sql查询
 * @param {string} strSql string sql code
 */

function queryString(strSql) {
    return new Promise((resolve, reject) => {
        connection.query(strSql, (erro, result, fields) => {
            if (erro) {
                reject(erro);
            } else {
                resolve(result);
            }
        });
    });
}

/**
 * sql查询，查询句带？
 * @param {string} strSql string sql code
 * @param {Array} args array of sql param
 */
function query(strSql, args) {
    if (typeof args === 'undefined' || args.length == 0) {
        return queryString(strSql);
    }

    return new Promise((resolve, reject) => {
        connection.query(strSql, args, (erro, result, fields) => {
            if (erro) {
                reject(erro);
            } else {
                resolve(result, fields);
            }
        });
    });
}

/**
 * sql查询，查询句带？
 * @param {string} strSql string sql code
 * @param {Array} args array of sql param
 */
async function sqlQuery(strSql, args) {
    try {
        let result = await query(strSql, args);
        return result;
    } catch (error) {
        return false;
    }
}

module.exports = sqlQuery;
